compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp

```diff
@@ -222,25 +244,133 @@ static bool setPipeliningMarkers(scf::ForOp forOp, bool pipelineStoreStage) {
   return copyToWorkgroupMemory;
 }
 
+/// Return true if the op is used as operand `index` of an mma.sync op.
+static bool isMMAOperand(Operation* op, int64_t index) {
+  OpOperand* use = &(*((op->getUses()).begin()));
+  if (auto extract = dyn_cast<vector::ExtractStridedSliceOp>(use->getOwner())) {
+    use = &(*((extract->getUses()).begin()));
+  }
+  if (!isa<nvgpu::MmaSyncOp>(use->getOwner())) return false;
+  return use->getOperandNumber() == index;
+}
+
+/// Return true if the op is used as operand A of an mma.sync op.
+static bool isAOperand(Operation* op) { return isMMAOperand(op, 0); }
+/// Return true if the op is used as operand B of an mma.sync op.
+static bool isBOperand(Operation* op) { return isMMAOperand(op, 1); }
+
+/// Return a pipelining schedule that gives good performance on Nvidia
+/// Ampere target.
+static void getNvidiaTensorCorePipeline(
+    scf::ForOp forOp, std::vector<std::pair<Operation*, unsigned>>& ops,
+    unsigned depth) {
+  bool loopCanBePipelined = false;
+  // TODO: Tune this and make it a more fine grain logic.
+  static constexpr int64_t prefetchSize = 4;
+  SmallVector<Operation*> stage0;
+  SmallVector<Operation*> stagePrefetch;
+  SmallVector<Operation*> stageLast;
+  int64_t sizeReadA = 0;
+  int64_t sizeReadB = 0;
+  for (Operation& op : forOp.getBody()->getOperations()) {
+    // Pipeline the most inner for op that should be a flat region.
+    if (op.getNumRegions() > 0) {
+      loopCanBePipelined = false;
+      break;
+    }
+    if (isa<gpu::BarrierOp, nvgpu::DeviceAsyncWaitOp>(op)) {
+      stagePrefetch.push_back(&op);
+    }
+    if (isa<nvgpu::MmaSyncOp>(op)) {
+      stageLast.push_back(&op);
+    }
+    if (isa<nvgpu::DeviceAsyncCopyOp, nvgpu::DeviceAsyncCreateGroupOp>(op)) {
+      stage0.push_back(&op);
```

Comment:
For consistency and understanding purposes, we have `stageLast`, `stage0`, and `stagePrefetch`. Will it read better if we have `stageLast`, `stagePrefetchGMEM`, and `stagePrefetchSMEM`